modules/core/shared/src/main/scala/lucuma/core/data/DisjointIntervalMap.scala

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) 2016-2025 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package lucuma.core.data
+
+import cats.collections.Diet
+import cats.collections.Discrete
+import cats.collections.Range
+import cats.kernel.CommutativeSemigroup
+import cats.kernel.Eq
+import cats.kernel.LowerBounded
+import cats.kernel.Order
+import cats.kernel.UpperBounded
+
+/** A map from keys to interval sets, where the interval sets are disjoint and non-empty. */
+sealed trait DisjointIntervalMap[K, V: Discrete: Order]:
```

Comment:
This seems like it could maybe replace or be used as basis of `TimeAccountingState` in the ODB. It's a map keyed by `TimestampInterval`.